consensus/src/consensus/inner.rs

```diff
@@ -0,0 +1,508 @@
+
+        // 3. Remove transactions from the mempool
+        for transaction in block.transactions.iter() {
+            self.memory_pool.remove(&transaction.id.into())?;
+        }
+
+        Ok(())
+    }
+
+    /// Generate a transaction by spending old records and specifying new record attributes
+    fn create_transaction(&self, request: CreateTransactionRequest) -> Result<TransactionResponse, ConsensusError> {
+        let mut rng = thread_rng();
+        // Offline execution to generate a DPC transaction
+        let transaction_kernel = <InstantiatedDPC as DPCScheme<DeserializedLedger<Components>>>::execute_offline(
+            self.public.public_parameters.system_parameters.clone(),
+            request
+                .old_records
+                .iter()
+                .map(|x| <DPCRecord<Components> as VMRecord>::deserialize(x))
+                .collect::<Result<Vec<_>, _>>()?,
+            request.old_account_private_keys.into_iter().map(|x| x.into()).collect(),
+            request.new_record_owners.into_iter().map(|x| x.into()).collect(),
+            &request.new_is_dummy_flags,
+            &request.new_values,
+            request
+                .new_payloads
+                .into_iter()
+                .map(|x| RecordPayload::from_bytes(&x[..]))
+                .collect(),
+            request.new_birth_program_ids,
+            request.new_death_program_ids,
+            request.memo,
+            self.public.parameters.network_id.id(),
+            &mut rng,
+        )?;
+
+        // Construct the program proofs
+        let (old_death_program_proofs, new_birth_program_proofs) = ConsensusParameters::generate_program_proofs(
+            &self.public.public_parameters,
+            &transaction_kernel,
+            &mut rng,
+        )?;
+
+        // Online execution to generate a DPC transaction
+        let (new_records, transaction) = InstantiatedDPC::execute_online(
+            &self.public.public_parameters,
+            transaction_kernel,
+            old_death_program_proofs,
+            new_birth_program_proofs,
+            &self.ledger.deserialize::<Components>(),
+            &mut rng,
+        )?;
+
+        let serialized_records = new_records.iter().map(|x| x.serialize()).collect::<Result<Vec<_>>>()?;
+
+        Ok(TransactionResponse {
+            records: serialized_records,
+            transaction: transaction.serialize()?,
+        })
+    }
+
+    /// Generate a transaction by spending old records and specifying new record attributes
+    fn create_partial_transaction(
+        &self,
+        request: CreatePartialTransactionRequest,
+    ) -> Result<TransactionResponse, ConsensusError> {
+        let mut rng = thread_rng();
+        // Offline execution to generate a DPC transaction
+        let transaction_kernel: Box<TransactionKernel<Components>> = request
+            .kernel
+            .downcast()
+            .expect("illegal kernel passed to create partial transaction");
```

Comment:
Same point here as with the other potential downcast panics.

Reply:
ditto pre-validated, safe but short/mid-term hacks